File: frontend/src/api/track.ts
/**
 * 赛道管理 API
 */
import request from './index'

export interface Track {
  id: string
  name: string
  display_name: string
  description: string | null
  sort_order: number
  is_active: number
  stock_count: number
  stocks: TrackStock[]
  created_at: string
}

export interface TrackStock {
  code: string
  name: string
  ipo_date: string | null
  status: string | null
}

export interface LabelDataPoint {
  stock_code: string
  trade_date: string
  close_px: number
  open_px: number
  high_px: number
  low_px: number
  volume: number
  amount: number
  fwd_1d_return: number | null
  fwd_5d_return: number | null
  fwd_20d_return: number | null
}

export function listTracks() {
  return request.get('/track/tracks')
}

export function getTrack(id: string) {
  return request.get(`/track/tracks/${id}`)
}

export function createTrack(data: { name: string; display_name: string; description?: string }) {
  return request.post('/track/tracks', data)
}

export function deleteTrack(id: string) {
  return request.delete(`/track/tracks/${id}`)
}

export function listStocks(trackId?: string) {
  const params = trackId ? { track_id: trackId } : {}
  return request.get('/track/stocks', { params })
}

export function assignStock(data: { stock_code: string; stock_name: string; track_ids: string[] }) {
  return request.post('/track/stocks/assign', data)
}

export function getLabels(stockCode: string) {
  return request.get(`/track/labels/${stockCode}`)
}

export function getWhitelist() {
  return request.get('/track/factors/whitelist')
}

export function getBlacklist() {
  return request.get('/track/factors/blacklist')
}

export function getTrackScore(trackName: string) {
  return request.get(`/ml/score/${trackName}`)
}

export function getAllTrackScores() {
  return request.get('/ml/scores')
}

export function trainTrackModel(trackName: string, params?: {
  num_leaves?: number
  max_depth?: number
  learning_rate?: number
  n_estimators?: number
  reg_alpha?: number
  reg_lambda?: number
  feature_fraction?: number
  bagging_fraction?: number
  min_child_samples?: number
}) {
  return request.post(`/ml/train/${trackName}`, params || {})
}

export function trainAllModels() {
  return request.get('/ml/train/all')
}

export function listTrackModels(trackName: string) {
  return request.get(`/ml/models/${trackName}`)
}

export function getScoreHistory(trackName: string, limit = 5) {
  return request.get(`/ml/score/history/${trackName}`, { params: { limit } })
}

export function getAllModels() {
  return request.get('/ml/models/all')
}

export function getFactorData(params?: { track_name?: string; stock_code?: string; max_rows?: number }) {
  return request.get('/ml/factors/data', { params })
}

export function getPortfolioSummary() {
  return request.get('/portfolio/summary')
}

export function getBacktestReport() {
  return request.get('/backtest/report')
}

export function getBacktestEquity() {
  return request.get('/backtest/equity')
}

// ── 策略回测 ──

export function listStrategies() {
  return request.get('/backtest/strategies')
}

export function getStrategyDetail(name: string) {
  return request.get(`/backtest/strategies/${name}`)
}

export function runStrategyBacktest(strategyName: string) {
  return request.post(`/backtest/strategy/${strategyName}`)
}

export function runSingleStockBacktest(stockCode: string, strategy?: string, lookback?: number, stopLoss?: number, useAi?: boolean) {
  const params: Record<string, any> = { strategy, lookback, stop_loss: stopLoss }
  if (useAi) params.use_ai = true
  return request.post(`/backtest/single/${stockCode}`, null, { params })
}

export function runBacktest(data: {
  initial_capital: number
  top_n: number
  rebalance_freq: string
  max_single_stock: number
  max_single_track: number
}) {
  return request.post('/backtest/run', data)
}

export function computeFeatures() {
  return request.post('/ml/compute-features')
}

export function screenFactors() {
  return request.post('/ml/screen-factors')
}

export function preprocessFeatures() {
  return request.post('/ml/preprocess')
}

export function runPipeline(step: string = 'all') {
  return request.post('/ml/run-pipeline', null, { params: { step } })
}

// ── 流水线运行日志 ──

export interface PipelineRun {
  id: number
  run_type: 'train' | 'backtest'
  status: string
  params_snapshot: Record<string, any> | null
  results_summary: Record<string, any> | null
  git_commit_hash: string | null
  feature_count: number | null
  duration_seconds: number | null
  created_at: string
}

export function getPipelineRuns(limit = 10, runType?: string) {
  const params: Record<string, any> = { limit }
  if (runType) params.run_type = runType
  return request.get('/ml/pipeline-runs', { params })
}

// ── 特征配置 CRUD ──

export interface FeatureConfig {
  id: number
  feature_name: string
  display_name: string
  category: string | null
  description: string | null
  formula: string | null
  interpretation: string | null
  default_params: Record<string, any> | null
  is_enabled: number
  is_user_defined: number
  created_at: string
  updated_at: string | null
}

export interface FeatureConfigCreate {
  feature_name: string
  display_name?: string
  category?: string
  description?: string
  formula?: string
  interpretation?: string
  default_params?: Record<string, any>
  is_enabled?: number
}

export interface FeatureConfigUpdate {
  display_name?: string
  category?: string
  description?: string
  formula?: string
  interpretation?: string
  default_params?: Record<string, any>
  is_enabled?: number
}

export function listFeatureConfigs(params?: { category?: string; enabled_only?: boolean }) {
  return request.get('/features/configs', { params })
}

export function getFeatureConfig(id: number) {
  return request.get(`/features/configs/${id}`)
}

export function createFeatureConfig(data: FeatureConfigCreate) {
  return request.post('/features/configs', data)
}

export function updateFeatureConfig(id: number, data: FeatureConfigUpdate) {
  return request.put(`/features/configs/${id}`, data)
}

export function deleteFeatureConfig(id: number) {
  return request.delete(`/features/configs/${id}`)
}

export function toggleFeatureConfig(id: number) {
  return request.put(`/features/configs/${id}/toggle`)
}

export function syncFeatureConfigs() {
  return request.post('/features/sync')
}

export function incrementalCompute() {
  return request.post('/features/compute/incremental')
}

// ── 量化交易员学习和成长平台 API ──

export function compareStrategies(data: {
  strategies: string[]
  track_name: string
  initial_capital?: number
  top_n?: number
  rebalance_freq?: string
  max_single_stock?: number
  max_single_track?: number
  stop_loss_pct?: number
  take_profit_pct?: number
}) {
  return request.post('/backtest/compare', data)
}

export function getBacktestHistory(limit = 20) {
  return request.get('/backtest/history', { params: { limit } })
}

export function getBacktestDetail(runId: number) {
  return request.get(`/backtest/history/${runId}`)
}

export function getLearningStats() {
  return request.get('/learning/stats')
}

File: backend/app/schemas/track.py
"""
赛道分类与标签系统 Pydantic Schema.
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# ── 赛道 ──────────────────────────────────────


class TrackCreate(BaseModel):
    name: str = Field(..., description="赛道标识名", examples=["semiconductor"])
    display_name: str = Field(..., description="赛道中文名", examples=["半导体"])
    description: str | None = None
    sort_order: int = 0


class TrackUpdate(BaseModel):
    display_name: str | None = None
    description: str | None = None
    sort_order: int | None = None
    is_active: int | None = None


class TrackStockInfo(BaseModel):
    code: str
    name: str
    ipo_date: str | None = None
    status: str | None = None

    model_config = ConfigDict(from_attributes=True)


class TrackResponse(BaseModel):
    id: str
    name: str
    display_name: str
    description: str | None = None
    sort_order: int
    is_active: int
    stock_count: int = 0
    stocks: list[TrackStockInfo] = []
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TrackListResponse(BaseModel):
    total: int
    items: list[TrackResponse]


# ── 自选股 ──────────────────────────────────────


class StockTrackAssign(BaseModel):
    stock_code: str = Field(..., description="股票代码", examples=["000636.SZ"])
    stock_name: str = Field(..., description="股票名称", examples=["风华高科"])
    track_ids: list[str] = Field(..., description="所属赛道 ID 列表")


class StockResponse(BaseModel):
    code: str
    name: str
    ipo_date: str | None = None
    status: str | None = None
    tracks: list[str] = []  # track names
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── 因子白名单 / 黑名单 ──────────────────────


class FactorWhiteListResponse(BaseModel):
    id: int
    factor_name: str
    factor_type: str
    category: str | None = None
    ic_mean: float = 0
    ir: float = 0
    rank_ic: float = 0
    lgb_importance: float = 0
    is_active: int = 1
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class FactorBlackListResponse(BaseModel):
    id: int
    factor_name: str
    reason: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── 数据标签 ──────────────────────────────────────


class LabeledDataPoint(BaseModel):
    """带标签的日线数据点"""
    stock_code: str
    trade_date: str
    open_px: float
    high_px: float
    low_px: float
    close_px: float
    volume: float
    amount: float
    is_stopped: bool
    is_limit_up: bool
    is_limit_down: bool
    # 未来收益标签（shift(1) 防止未来函数）
    fwd_1d_return: float | None = None
    fwd_5d_return: float | None = None
    fwd_20d_return: float | None = None
    fwd_1d_excess_return: float | None = None  # 超额收益（相对赛道中位数）
    fwd_5d_excess_return: float | None = None
    fwd_20d_excess_return: float | None = None


class LabeledDataResponse(BaseModel):
    stock_code: str
    stock_name: str
    track_names: list[str]
    data_points: list[LabeledDataPoint]
    total: int


# ── 回测参数 ──────────────────────────────────────


class TrainModelParams(BaseModel):
    """训练模型可配置参数（可选覆盖，不传则用默认值）"""
    num_leaves: int | None = Field(default=None, description="叶子节点数", ge=4, le=128)
    max_depth: int | None = Field(default=None, description="最大深度", ge=3, le=20)
    learning_rate: float | None = Field(default=None, description="学习率", gt=0, le=0.5)
    n_estimators: int | None = Field(default=None, description="迭代轮数", ge=100, le=5000)
    reg_alpha: float | None = Field(default=None, description="L1 正则", ge=0, le=10)
    reg_lambda: float | None = Field(default=None, description="L2 正则", ge=0, le=10)
    feature_fraction: float | None = Field(default=None, description="特征采样比例", gt=0, le=1)
    bagging_fraction: float | None = Field(default=None, description="样本采样比例", gt=0, le=1)
    min_child_samples: int | None = Field(default=None, description="叶节点最小样本", ge=5, le=200)


class ScoreHistoryResponse(BaseModel):
    """评分历史记录"""
    id: int
    track_name: str
    model_id: str
    stock_code: str
    stock_name: str | None = None
    score: float = 0
    rank: int = 0
    train_r2: float = 0
    val_r2: float = 0
    test_r2: float = 0
    params_snapshot: dict | None = None
    scored_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BacktestRunParams(BaseModel):
    """人工回测可配置参数"""
    initial_capital: int = Field(default=100000, description="初始资金", ge=10000)
    top_n: int = Field(default=3, description="每赛道买入 Top-N 只", ge=1, le=10)
    rebalance_freq: str = Field(default="W", description="调仓频率: W=周频, M=月频")
    max_single_stock: float = Field(default=0.20, description="单票上限", ge=0.05, le=0.50)
    max_single_track: float = Field(default=0.50, description="单赛道上限", ge=0.10, le=0.80)


class PipelineRunResponse(BaseModel):
    """流水线运行记录"""
    id: int
    run_type: str
    status: str
    params_snapshot: dict | None = None
    results_summary: dict | None = None
    git_commit_hash: str | None = None
    feature_count: int | None = None
    duration_seconds: float | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)
